### ISSUE TITLE
Writing to /dev/stdout from a forked process does not work

### ISSUE THREAD

**Issue by sleighter** [flynn/flynn#3589]
Related to #3535. Opening a new issue since that one was closed with a fix, but this issue persists.

We're using the heroku-ruby buildpack and the PDFKit and wkhtmltopdf-binary gems. We are running flynn-host nightly v20161024.0, so we have the fix from #3535. 

PDFKit works by forking a child process and collecting the child process's stdout. At this time, the child process's output appears in the flynn logs, but is not available to the ruby app that originally forked the process. So we're seeing the PDF binary in our logs, but our application raises an exception when the forked process completes but the app doesn't see any output.

**Comment by sleighter** [comment]
I implemented a workaround by writing the PDFs to a tempfile, then reading the contents of that file, rather than attempting to consume the forked process's output directly. However, this issue would likely affect any application or library that relies the output of a forked process.

**Comment by lmars** [comment]
To clarify the issue:

* PDFKit runs `wkhtmltopdf` with `-` as the output, which leads to the output being written to the `/dev/stdout` file (see [here](https://github.com/wkhtmltopdf/wkhtmltopdf/blob/master/src/lib/pdfconverter.cc#L335-L342))
* Usually `/dev/stdout` is a symlink to `/proc/self/fd/1` (so writing to it causes the kernel to write to the process's fd 1), but Flynn creates `/dev/stdout` as a FIFO connected to the container's stdout log because fd 1 is a socket file which cannot be opened with `open(2)` (see #3371)

To demonstrate this issue, running a subshell which writes to `/dev/stdout` and capturing into a variable should work as follows:

```
$ FOO="$(echo foo > /dev/stdout)"; echo "FOO has value \"${FOO}\""
FOO has value "foo"
```

Running this in Flynn does not capture the output into the variable, instead outputs directly to the container's stdout log:

```
$ flynn run bash <<< 'FOO="$(echo foo > /dev/stdout)"; echo "FOO has value \"${FOO}\""'
foo
FOO has value ""
```

Not round-tripping via `/dev/stdout` does work as expected though:

```
$ flynn run bash <<< 'FOO="$(echo foo)"; echo "FOO has value \"${FOO}\""'
FOO has value "foo"
```

As does writing to `/proc/self/fd/1`:

```
$ flynn run bash <<< 'FOO="$(echo foo > /proc/self/fd/1)"; echo "FOO has value \"${FOO}\""'
FOO has value "foo"
```

We could fix this by implementing my original patch from #3371 (i.e. create a FIFO for the container's stdout and have containerinit copy it to the socket for all jobs) or control this behaviour with a flag (although due to the subtlety of this bug, users probably won't realise they need to set the flag).

Open to suggestions.

**Comment by lmars** [comment]
Also, in the specific case of PDFKit, a valid workaround is to round-trip through a temporary file as @sleighter mentions:

```ruby
pdf = Dir.mktmpdir { |d| PDFKit.new(html).to_pdf("#{d}/pdf"); File.read("#{d}/pdf") }
```

**Comment by titanous** [comment]
@lmars I think we'll have to go the FIFO route. Will `splice(2)` still work?

**Comment by lmars** [comment]
@sleighter we've made another patch which fixes this (#3592):


Before (note the `RuntimeError`):

```
$ flynn run ruby -r pdfkit <<< 'puts PDFKit.new("<html></html>").to_pdf' | head -5
%PDF-1.4
1 0 obj
<<
/Title (??)
/Creator (??wkhtmltopdf 0.12.3)
/app/vendor/bundle/ruby/2.3.0/gems/pdfkit-0.8.2/lib/pdfkit/pdfkit.rb:71:in `to_pdf': command failed (exitstatus=0): /app/vendor/bundle/bin/wkhtmltopdf --quiet --page-size Letter --margin-top 0.75in --margin-right 0.75in --margin-bottom 0.75in --margin-left 0.75in --encoding UTF-8 - - (RuntimeError)
        from -:1:in `<main>'
```

With the patch:

```
$ flynn run ruby -r pdfkit <<< 'puts PDFKit.new("<html></html>").to_pdf' | head -5
%PDF-1.4
1 0 obj
<<
/Title (??)
/Creator (??wkhtmltopdf 0.12.3)
```
